File: packages/esm_format/src/esm_format/serialize.py
"""
ESM Format serialization module.

This module provides functions to serialize ESM format objects to JSON,
with optional file writing capability.
"""

import json
from pathlib import Path
from typing import Union, Dict, Any, Optional

from .esm_types import (
    EsmFile, Metadata, Model, ReactionSystem, ModelVariable, Equation,
    Species, Parameter, Reaction, ExprNode, Expr, AffectEquation,
    ContinuousEvent, DiscreteEvent, DiscreteEventTrigger, FunctionalAffect,
    DataLoader, DataLoaderType, Operator,
    CouplingEntry, CouplingType, Domain, Solver, SolverType,
    OperatorComposeCoupling, Couple2Coupling, VariableMapCoupling,
    OperatorApplyCoupling, CallbackCoupling, EventCoupling,
    Reference, TemporalDomain, SpatialDimension, CoordinateTransform,
    InitialCondition, BoundaryCondition
)


def _serialize_expression(expr: Expr) -> Union[int, float, str, Dict[str, Any]]:
    """Serialize an expression to JSON-compatible format."""
    if isinstance(expr, (int, float, str)):
        return expr
    elif isinstance(expr, ExprNode):
        result = {
            "op": expr.op,
            "args": [_serialize_expression(arg) for arg in expr.args]
        }
        if expr.wrt is not None:
            result["wrt"] = expr.wrt
        if expr.dim is not None:
            result["dim"] = expr.dim
        return result
    else:
        raise ValueError(f"Invalid expression type: {type(expr)}")


def _serialize_equation(equation: Equation) -> Dict[str, Any]:
    """Serialize an equation to JSON-compatible format."""
    result = {
        "lhs": _serialize_expression(equation.lhs),
        "rhs": _serialize_expression(equation.rhs)
    }
    if equation._comment is not None:
        result["_comment"] = equation._comment
    return result


def _serialize_affect_equation(affect: AffectEquation) -> Dict[str, Any]:
    """Serialize an affect equation to JSON-compatible format."""
    return {
        "lhs": affect.lhs,
        "rhs": _serialize_expression(affect.rhs)
    }


def _serialize_model_variable(variable: ModelVariable) -> Dict[str, Any]:
    """Serialize a model variable to JSON-compatible format."""
    result = {
        "type": variable.type
    }
    if variable.units is not None:
        result["units"] = variable.units
    if variable.default is not None:
        result["default"] = variable.default
    if variable.description is not None:
        result["description"] = variable.description
    if variable.expression is not None:
        result["expression"] = _serialize_expression(variable.expression)
    return result


def _serialize_discrete_event_trigger(trigger: DiscreteEventTrigger) -> Dict[str, Any]:
    """Serialize a discrete event trigger to JSON-compatible format."""
    result = {"type": trigger.type}

    if trigger.type == "condition":
        result["expression"] = _serialize_expression(trigger.value)
    elif trigger.type == "periodic":
        result["interval"] = trigger.value
    elif trigger.type == "preset_times":
        result["times"] = trigger.value

    return result


def _serialize_continuous_event(event: ContinuousEvent) -> Dict[str, Any]:
    """Serialize a continuous event to JSON-compatible format."""
    result = {
        "conditions": [_serialize_expression(cond) for cond in event.conditions],  # Fixed: use plural conditions
        "affects": [_serialize_affect_equation(affect) for affect in event.affects]
    }
    if event.name:
        result["name"] = event.name
    if event.priority != 0:
        result["priority"] = event.priority

    # Add new fields
    if event.affect_neg is not None:
        result["affect_neg"] = [_serialize_affect_equation(affect) for affect in event.affect_neg]
    if event.root_find and event.root_find != 'left':  # Only include if not default
        result["root_find"] = event.root_find
    if event.reinitialize:  # Only include if True (not default False)
        result["reinitialize"] = event.reinitialize
    if event.description:
        result["description"] = event.description

    return result


def _serialize_discrete_event(event: DiscreteEvent) -> Dict[str, Any]:
    """Serialize a discrete event to JSON-compatible format."""
    result = {
        "trigger": _serialize_discrete_event_trigger(event.trigger),
        "affects": [_serialize_affect_equation(affect) for affect in event.affects]
    }
    if event.name:
        result["name"] = event.name
    if event.priority != 0:
        result["priority"] = event.priority

    return result


def _serialize_model(model: Model) -> Dict[str, Any]:
    """Serialize a model to JSON-compatible format."""
    result = {}

    # Serialize variables (required by schema)
    result["variables"] = {}
    if model.variables:
        result["variables"] = {
            name: _serialize_model_variable(var)
            for name, var in model.variables.items()
        }

    # Serialize equations (required by schema)
    result["equations"] = []
    if model.equations:
        result["equations"] = [_serialize_equation(eq) for eq in model.equations]

    return result


def _serialize_species(species: Species) -> Dict[str, Any]:
    """Serialize a species to JSON-compatible format."""
    result = {}
    if species.units is not None:
        result["units"] = species.units
    if species.description is not None:
        result["description"] = species.description
    return result


def _serialize_parameter(parameter: Parameter) -> Dict[str, Any]:
    """Serialize a parameter to JSON-compatible format."""
    result = {}
    if parameter.units is not None:
        result["units"] = parameter.units
    if parameter.description is not None:
        result["description"] = parameter.description
    if isinstance(parameter.value, (int, float)):
        result["default"] = parameter.value
    return result


def _serialize_reaction(reaction: Reaction) -> Dict[str, Any]:
    """Serialize a reaction to JSON-compatible format."""
    result = {
        "id": reaction.name
    }

    if reaction.name:
        result["name"] = reaction.name

    # Serialize substrates (reactants)
    if reaction.reactants:
        result["substrates"] = [
            {"species": species, "stoichiometry": int(coeff)}
            for species, coeff in reaction.reactants.items()
        ]
    else:
        result["substrates"] = None

    # Serialize products
    if reaction.products:
        result["products"] = [
            {"species": species, "stoichiometry": int(coeff)}
            for species, coeff in reaction.products.items()
        ]
    else:
        result["products"] = None

    # Serialize rate
    if reaction.rate_constant is not None:
        result["rate"] = _serialize_expression(reaction.rate_constant)

    return result


def _serialize_reaction_system(rs: ReactionSystem) -> Dict[str, Any]:
    """Serialize a reaction system to JSON-compatible format."""
    result = {}

    # Serialize species
    if rs.species:
        result["species"] = {
            sp.name: _serialize_species(sp)
            for sp in rs.species
        }
    else:
        result["species"] = {}

    # Serialize parameters
    if rs.parameters:
        result["parameters"] = {
            param.name: _serialize_parameter(param)
            for param in rs.parameters
        }
    else:
        result["parameters"] = {}

    # Serialize reactions
    if rs.reactions:
        result["reactions"] = [_serialize_reaction(reaction) for reaction in rs.reactions]
    else:
        result["reactions"] = []

    return result


def _serialize_reference(reference: Reference) -> Dict[str, Any]:
    """Serialize a reference to JSON-compatible format."""
    result = {}
    if reference.title:
        result["citation"] = reference.title
    if reference.doi is not None:
        result["doi"] = reference.doi
    if reference.url is not None:
        result["url"] = reference.url
    return result


def _serialize_metadata(metadata: Metadata) -> Dict[str, Any]:
    """Serialize metadata to JSON-compatible format."""
    result = {
        "name": metadata.title
    }

    if metadata.description is not None:
        result["description"] = metadata.description
    if metadata.authors:
        result["authors"] = metadata.authors
    if metadata.created is not None:
        result["created"] = metadata.created
    if metadata.modified is not None:
        result["modified"] = metadata.modified
    if metadata.keywords:
        result["tags"] = metadata.keywords
    if metadata.references:
        result["references"] = [_serialize_reference(ref) for ref in metadata.references]

    return result


def _serialize_domain(domain: Domain) -> Dict[str, Any]:
    """Serialize a domain to JSON-compatible format."""
    result = {}

    if domain.independent_variable:
        result["independent_variable"] = domain.independent_variable

    # Serialize temporal domain
    if domain.temporal:
        temporal_data = {
            "start": domain.temporal.start,
            "end": domain.temporal.end
        }
        if domain.temporal.reference_time:
            temporal_data["reference_time"] = domain.temporal.reference_time
        result["temporal"] = temporal_data

    # Serialize spatial domain
    if domain.spatial:
        spatial_data = {}
        for dim_name, dim_spec in domain.spatial.items():
            dim_data = {
                "min": dim_spec.min,
                "max": dim_spec.max,
                "units": dim_spec.units
            }
            if dim_spec.grid_spacing is not None:
                dim_data["grid_spacing"] = dim_spec.grid_spacing
            spatial_data[dim_name] = dim_data
        result["spatial"] = spatial_data

    # Serialize coordinate transforms
    if domain.coordinate_transforms:
        result["coordinate_transforms"] = [
            {
                "id": transform.id,
                "description": transform.description,
                "dimensions": transform.dimensions
            }
            for transform in domain.coordinate_transforms
        ]

    # Serialize spatial reference
    if domain.spatial_ref:
        result["spatial_ref"] = domain.spatial_ref

    # Serialize initial conditions
    if domain.initial_conditions:
        ic = domain.initial_conditions
        ic_data = {"type": ic.type.value}

        if ic.value is not None:
            ic_data["value"] = ic.value
        if ic.values is not None:
            ic_data["values"] = ic.values
        if ic.function is not None:
            ic_data["function"] = ic.function
        if ic.data_source is not None:
            ic_data["path"] = ic.data_source  # Schema uses "path" for data source

        result["initial_conditions"] = ic_data

    # Serialize boundary conditions
    if domain.boundary_conditions:
        bc_data = []
        for bc in domain.boundary_conditions:
            bc_dict = {
                "type": bc.type.value,
                "dimensions": bc.dimensions
            }
            if bc.value is not None:
                bc_dict["value"] = bc.value
            if bc.function is not None:
                bc_dict["function"] = bc.function
            bc_data.append(bc_dict)
        result["boundary_conditions"] = bc_data

    return result


def _serialize_data_loader(loader: DataLoader) -> Dict[str, Any]:
    """Serialize a data loader to JSON-compatible format."""
    result = {}

    # Map our enum directly to schema types (they now match)
    type_mapping = {
        DataLoaderType.GRIDDED_DATA: "gridded_data",
        DataLoaderType.EMISSIONS: "emissions",
        DataLoaderType.TIMESERIES: "timeseries",
        DataLoaderType.STATIC: "static",
        DataLoaderType.CALLBACK: "callback"
    }
    result["type"] = type_mapping.get(loader.type, "static")

    # Schema uses loader_id for source
    if loader.source:
        result["loader_id"] = loader.source

    # Schema uses config for format_options
    if loader.format_options:
        result["config"] = loader.format_options

    # Schema uses provides for variables
    if loader.variables:
        result["provides"] = {var: {} for var in loader.variables}

    return result


def _serialize_operator(operator: Operator) -> Dict[str, Any]:
    """Serialize an operator to JSON-compatible format."""
    result = {}

    # Schema requires operator_id
    result["operator_id"] = operator.operator_id

    # Schema requires needed_vars
    result["needed_vars"] = operator.needed_vars

    # Optional fields
    if operator.modifies is not None:
        result["modifies"] = operator.modifies

    if operator.config:
        result["config"] = operator.config

    if operator.description:
        result["description"] = operator.description

    if operator.reference:
        result["reference"] = _serialize_reference(operator.reference)

    return result


def _serialize_coupling_entry(coupling: CouplingEntry) -> Dict[str, Any]:
    """Serialize a coupling entry to JSON-compatible format."""
    result = {}

    # Add description if present
    if coupling.description:
        result["description"] = coupling.description

    # Handle different coupling types
    if isinstance(coupling, OperatorComposeCoupling):
        result["type"] = "operator_compose"
        if coupling.systems:
            result["systems"] = coupling.systems
        if coupling.translate:
            result["translate"] = coupling.translate

    elif isinstance(coupling, Couple2Coupling):
        result["type"] = "couple2"
        if coupling.systems:
            result["systems"] = coupling.systems
        if coupling.coupletype_pair:
            result["coupletype_pair"] = coupling.coupletype_pair
        if coupling.connector:
            result["connector"] = {
                "equations": [
                    {
                        "from": eq.from_var,
                        "to": eq.to_var,
                        "transform": eq.transform,
                        **({"expression": _serialize_expression(eq.expression)} if eq.expression else {})
                    }
                    for eq in coupling.connector.equations
                ]
            }

    elif isinstance(coupling, VariableMapCoupling):
        result["type"] = "variable_map"
        if coupling.from_var:
            result["from"] = coupling.from_var
        if coupling.to_var:
            result["to"] = coupling.to_var
        if coupling.transform:
            result["transform"] = coupling.transform
        if coupling.factor is not None:
            result["factor"] = coupling.factor

    elif isinstance(coupling, OperatorApplyCoupling):
        result["type"] = "operator_apply"
        if coupling.operator:
            result["operator"] = coupling.operator

    elif isinstance(coupling, CallbackCoupling):
        result["type"] = "callback"
        if coupling.callback_id:
            result["callback_id"] = coupling.callback_id
        if coupling.config:
            result["config"] = coupling.config

    elif isinstance(coupling, EventCoupling):
        result["type"] = "event"
        if coupling.event_type:
            result["event_type"] = coupling.event_type
        if coupling.conditions:
            result["conditions"] = [_serialize_expression(cond) for cond in coupling.conditions]
        if coupling.trigger:
            result["trigger"] = _serialize_discrete_event_trigger(coupling.trigger)
        if coupling.affects:
            result["affects"] = [_serialize_affect_equation(affect) for affect in coupling.affects]
        if coupling.affect_neg:
            result["affect_neg"] = [_serialize_affect_equation(affect) for affect in coupling.affect_neg]
        if coupling.discrete_parameters:
            result["discrete_parameters"] = coupling.discrete_parameters
        if coupling.root_find:
            result["root_find"] = coupling.root_find
        if coupling.reinitialize is not None:
            result["reinitialize"] = coupling.reinitialize

    return result


def _serialize_solver(solver: Solver) -> Dict[str, Any]:
    """Serialize a solver to JSON-compatible format."""
    result = {}

    # Schema uses strategy for algorithm
    if solver.algorithm:
        result["strategy"] = solver.algorithm

    # Schema uses config for parameters
    config = {}
    if solver.parameters:
        config.update(solver.parameters)

    # Add tolerances to config
    if solver.tolerances:
        stiff_kwargs = {}
        if "absolute" in solver.tolerances:
            stiff_kwargs["abstol"] = solver.tolerances["absolute"]
        if "relative" in solver.tolerances:
            stiff_kwargs["reltol"] = solver.tolerances["relative"]
        if stiff_kwargs:
            config["stiff_kwargs"] = stiff_kwargs

    if config:
        result["config"] = config

    return result


def _serialize_esm_file(esm_file: EsmFile) -> Dict[str, Any]:
    """Serialize an ESM file to JSON-compatible format."""
    result = {
        "esm": esm_file.version,
        "metadata": _serialize_metadata(esm_file.metadata)
    }

    # Serialize models
    if esm_file.models:
        result["models"] = {
            model_name: _serialize_model(model)
            for model_name, model in esm_file.models.items()
        }

    # Serialize reaction systems
    if esm_file.reaction_systems:
        result["reaction_systems"] = {
            rs_name: _serialize_reaction_system(rs)
            for rs_name, rs in esm_file.reaction_systems.items()
        }

    # Serialize domain
    if esm_file.domain:
        result["domain"] = _serialize_domain(esm_file.domain)

    # Serialize data loaders
    if esm_file.data_loaders:
        result["data_loaders"] = {
            loader.name: _serialize_data_loader(loader)
            for loader in esm_file.data_loaders
        }

    # Serialize operators
    if esm_file.operators:
        result["operators"] = {
            operator.name: _serialize_operator(operator)
            for operator in esm_file.operators
        }

    # Serialize coupling
    if esm_file.coupling:
        result["coupling"] = [
            _serialize_coupling_entry(coupling)
            for coupling in esm_file.coupling
        ]

    # Serialize solver
    if esm_file.solver:
        result["solver"] = _serialize_solver(esm_file.solver)

    # Serialize events
    if esm_file.events:
        # For now, serialize events as a simple list at top level
        # TODO: Properly distribute events back to their parent models/reaction_systems
        continuous_events = []
        discrete_events = []

        for event in esm_file.events:
            if isinstance(event, ContinuousEvent):
                continuous_events.append(_serialize_continuous_event(event))
            elif isinstance(event, DiscreteEvent):
                discrete_events.append(_serialize_discrete_event(event))

        if continuous_events:
            result["continuous_events"] = continuous_events
        if discrete_events:
            result["discrete_events"] = discrete_events

    return result


def save(esm_file: EsmFile, path: Optional[Union[str, Path]] = None) -> str:
    """
    Serialize an ESM file to JSON string, optionally writing to file.

    Args:
        esm_file: The EsmFile object to serialize
        path: Optional file path to write the JSON to

    Returns:
        JSON string representation of the ESM file

    Raises:
        IOError: If writing to file fails
    """
    # Serialize to dictionary
    data = _serialize_esm_file(esm_file)

    # Convert to JSON string with nice formatting
    json_str = json.dumps(data, indent=2, ensure_ascii=False)

    # Write to file if path provided
    if path is not None:
        with open(path, 'w') as f:
            f.write(json_str)

    return json_str

File: packages/ESMFormat.jl/src/serialize.jl
"""
ESM Format JSON Serialization

Provides functionality to serialize EsmFile objects to JSON strings.
"""

using JSON3

"""
    serialize_expression(expr::Expr) -> Any

Serialize an Expression to JSON-compatible format.
Handles the union type discrimination.
"""
function serialize_expression(expr::Expr)
    if isa(expr, NumExpr)
        return expr.value
    elseif isa(expr, VarExpr)
        return expr.name
    elseif isa(expr, OpExpr)
        result = Dict{String,Any}(
            "op" => expr.op,
            "args" => [serialize_expression(arg) for arg in expr.args]
        )
        if expr.wrt !== nothing
            result["wrt"] = expr.wrt
        end
        if expr.dim !== nothing
            result["dim"] = expr.dim
        end
        return result
    else
        throw(ArgumentError("Unknown expression type: $(typeof(expr))"))
    end
end

"""
    serialize_model_variable_type(var_type::ModelVariableType) -> String

Serialize ModelVariableType enum to string.
"""
function serialize_model_variable_type(var_type::ModelVariableType)::String
    if var_type == StateVariable
        return "state"
    elseif var_type == ParameterVariable
        return "parameter"
    elseif var_type == ObservedVariable
        return "observed"
    else
        throw(ArgumentError("Unknown ModelVariableType: $(var_type)"))
    end
end

"""
    serialize_trigger(trigger::DiscreteEventTrigger) -> Dict{String,Any}

Serialize DiscreteEventTrigger to JSON-compatible format.
"""
function serialize_trigger(trigger::DiscreteEventTrigger)::Dict{String,Any}
    if isa(trigger, ConditionTrigger)
        return Dict("expression" => serialize_expression(trigger.expression))
    elseif isa(trigger, PeriodicTrigger)
        result = Dict("period" => trigger.period)
        if trigger.phase != 0.0
            result["phase"] = trigger.phase
        end
        return result
    elseif isa(trigger, PresetTimesTrigger)
        return Dict("times" => trigger.times)
    else
        throw(ArgumentError("Unknown DiscreteEventTrigger type: $(typeof(trigger))"))
    end
end

"""
    serialize_event(event::EventType) -> Dict{String,Any}

Serialize EventType to JSON-compatible format.
"""
function serialize_event(event::EventType)::Dict{String,Any}
    if isa(event, ContinuousEvent)
        result = Dict{String,Any}(
            "conditions" => [serialize_expression(c) for c in event.conditions],
            "affects" => [serialize_affect_equation(a) for a in event.affects]
        )
        if event.description !== nothing
            result["description"] = event.description
        end
        return result
    elseif isa(event, DiscreteEvent)
        result = Dict{String,Any}(
            "trigger" => serialize_trigger(event.trigger),
            "affects" => [serialize_functional_affect(a) for a in event.affects]
        )
        if event.description !== nothing
            result["description"] = event.description
        end
        return result
    else
        throw(ArgumentError("Unknown EventType: $(typeof(event))"))
    end
end

"""
    serialize_discrete_event(event::DiscreteEvent) -> Dict{String,Any}

Serialize DiscreteEvent to JSON-compatible format.
"""
function serialize_discrete_event(event::DiscreteEvent)::Dict{String,Any}
    result = Dict{String,Any}(
        "trigger" => serialize_trigger(event.trigger),
        "affects" => [serialize_functional_affect(a) for a in event.affects]
    )
    if event.description !== nothing
        result["description"] = event.description
    end
    return result
end

"""
    serialize_continuous_event(event::ContinuousEvent) -> Dict{String,Any}

Serialize ContinuousEvent to JSON-compatible format.
"""
function serialize_continuous_event(event::ContinuousEvent)::Dict{String,Any}
    result = Dict{String,Any}(
        "conditions" => [serialize_expression(c) for c in event.conditions],
        "affects" => [serialize_affect_equation(a) for a in event.affects]
    )
    if event.description !== nothing
        result["description"] = event.description
    end
    return result
end

"""
    serialize_affect_equation(affect::AffectEquation) -> Dict{String,Any}

Serialize AffectEquation to JSON-compatible format.
"""
function serialize_affect_equation(affect::AffectEquation)::Dict{String,Any}
    return Dict{String,Any}(
        "lhs" => affect.lhs,
        "rhs" => serialize_expression(affect.rhs)
    )
end

"""
    serialize_functional_affect(affect::FunctionalAffect) -> Dict{String,Any}

Serialize FunctionalAffect to JSON-compatible format.
"""
function serialize_functional_affect(affect::FunctionalAffect)::Dict{String,Any}
    result = Dict{String,Any}(
        "target" => affect.target,
        "expression" => serialize_expression(affect.expression)
    )
    if affect.operation != "set"
        result["operation"] = affect.operation
    end
    return result
end

"""
    serialize_model_variable(var::ModelVariable) -> Dict{String,Any}

Serialize ModelVariable to JSON-compatible format.
"""
function serialize_model_variable(var::ModelVariable)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => serialize_model_variable_type(var.type)
    )
    if var.default !== nothing
        result["default"] = var.default
    end
    if var.description !== nothing
        result["description"] = var.description
    end
    if var.expression !== nothing
        result["expression"] = serialize_expression(var.expression)
    end
    return result
end

"""
    serialize_equation(eq::Equation) -> Dict{String,Any}

Serialize Equation to JSON-compatible format.
"""
function serialize_equation(eq::Equation)::Dict{String,Any}
    return Dict{String,Any}(
        "lhs" => serialize_expression(eq.lhs),
        "rhs" => serialize_expression(eq.rhs)
    )
end

"""
    serialize_model(model::Model) -> Dict{String,Any}

Serialize Model to JSON-compatible format.
"""
function serialize_model(model::Model)::Dict{String,Any}
    result = Dict{String,Any}(
        "variables" => Dict(k => serialize_model_variable(v) for (k, v) in model.variables),
        "equations" => [serialize_equation(eq) for eq in model.equations]
    )

    # Serialize discrete events if present
    if !isempty(model.discrete_events)
        result["discrete_events"] = [serialize_discrete_event(ev) for ev in model.discrete_events]
    end

    # Serialize continuous events if present
    if !isempty(model.continuous_events)
        result["continuous_events"] = [serialize_continuous_event(ev) for ev in model.continuous_events]
    end

    # Add subsystems if present
    if !isempty(model.subsystems)
        result["subsystems"] = Dict(k => serialize_model(v) for (k, v) in model.subsystems)
    end

    return result
end

"""
    serialize_species(species::Species) -> Dict{String,Any}

Serialize Species to JSON-compatible format.
"""
function serialize_species(species::Species)::Dict{String,Any}
    result = Dict{String,Any}("name" => species.name)
    if species.units !== nothing
        result["units"] = species.units
    end
    if species.default !== nothing
        result["default"] = species.default
    end
    if species.description !== nothing
        result["description"] = species.description
    end
    return result
end

"""
    serialize_parameter(param::Parameter) -> Dict{String,Any}

Serialize Parameter to JSON-compatible format.
"""
function serialize_parameter(param::Parameter)::Dict{String,Any}
    result = Dict{String,Any}(
        "name" => param.name,
        "default" => param.default
    )
    if param.description !== nothing
        result["description"] = param.description
    end
    if param.units !== nothing
        result["units"] = param.units
    end
    return result
end

"""
    serialize_reaction(reaction::Reaction) -> Dict{String,Any}

Serialize Reaction to JSON-compatible format.
"""
function serialize_reaction(reaction::Reaction)::Dict{String,Any}
    result = Dict{String,Any}(
        "id" => reaction.id,
        "rate" => serialize_expression(reaction.rate)
    )

    if reaction.name !== nothing
        result["name"] = reaction.name
    end

    # Handle substrates (can be null for source reactions)
    if reaction.substrates !== nothing
        result["substrates"] = [
            Dict("species" => entry.species, "stoichiometry" => entry.stoichiometry)
            for entry in reaction.substrates
        ]
    else
        result["substrates"] = nothing
    end

    # Handle products (can be null for sink reactions)
    if reaction.products !== nothing
        result["products"] = [
            Dict("species" => entry.species, "stoichiometry" => entry.stoichiometry)
            for entry in reaction.products
        ]
    else
        result["products"] = nothing
    end

    if reaction.reference !== nothing
        result["reference"] = serialize_reference(reaction.reference)
    end

    return result
end

"""
    serialize_reaction_system(rs::ReactionSystem) -> Dict{String,Any}

Serialize ReactionSystem to JSON-compatible format.
"""
function serialize_reaction_system(rs::ReactionSystem)::Dict{String,Any}
    result = Dict{String,Any}(
        "species" => [serialize_species(s) for s in rs.species],
        "reactions" => [serialize_reaction(r) for r in rs.reactions]
    )
    if !isempty(rs.parameters)
        result["parameters"] = [serialize_parameter(p) for p in rs.parameters]
    end
    return result
end

"""
    serialize_data_loader(loader::DataLoader) -> Dict{String,Any}

Serialize DataLoader to JSON-compatible format.
"""
function serialize_data_loader(loader::DataLoader)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => loader.type,
        "source" => loader.source
    )
    if !isempty(loader.parameters)
        result["parameters"] = loader.parameters
    end
    if loader.description !== nothing
        result["description"] = loader.description
    end
    return result
end

"""
    serialize_operator(op::Operator) -> Dict{String,Any}

Serialize Operator to JSON-compatible format.
"""
function serialize_operator(op::Operator)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => op.type,
        "name" => op.name
    )
    if !isempty(op.parameters)
        result["parameters"] = op.parameters
    end
    if op.description !== nothing
        result["description"] = op.description
    end
    return result
end

"""
    serialize_coupling_entry(entry::CouplingEntry) -> Dict{String,Any}

Serialize CouplingEntry to JSON-compatible format based on concrete type.
"""
function serialize_coupling_entry(entry::CouplingEntry)::Dict{String,Any}
    if entry isa CouplingOperatorCompose
        return serialize_operator_compose(entry)
    elseif entry isa CouplingCouple2
        return serialize_couple2(entry)
    elseif entry isa CouplingVariableMap
        return serialize_variable_map(entry)
    elseif entry isa CouplingOperatorApply
        return serialize_operator_apply(entry)
    elseif entry isa CouplingCallback
        return serialize_callback(entry)
    elseif entry isa CouplingEvent
        return serialize_event(entry)
    else
        throw(ArgumentError("Unknown CouplingEntry type: $(typeof(entry))"))
    end
end

"""
    serialize_operator_compose(entry::CouplingOperatorCompose) -> Dict{String,Any}

Serialize operator_compose coupling entry.
"""
function serialize_operator_compose(entry::CouplingOperatorCompose)::Dict{String,Any}
    result = Dict{String,Any}("type" => "operator_compose", "systems" => entry.systems)

    if entry.translate !== nothing
        result["translate"] = entry.translate
    end
    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_couple2(entry::CouplingCouple2) -> Dict{String,Any}

Serialize couple2 coupling entry.
"""
function serialize_couple2(entry::CouplingCouple2)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => "couple2",
        "systems" => entry.systems,
        "coupletype_pair" => entry.coupletype_pair,
        "connector" => entry.connector
    )

    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_variable_map(entry::CouplingVariableMap) -> Dict{String,Any}

Serialize variable_map coupling entry.
"""
function serialize_variable_map(entry::CouplingVariableMap)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => "variable_map",
        "from" => entry.from,
        "to" => entry.to,
        "transform" => entry.transform
    )

    if entry.factor !== nothing
        result["factor"] = entry.factor
    end
    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_operator_apply(entry::CouplingOperatorApply) -> Dict{String,Any}

Serialize operator_apply coupling entry.
"""
function serialize_operator_apply(entry::CouplingOperatorApply)::Dict{String,Any}
    result = Dict{String,Any}("type" => "operator_apply", "operator" => entry.operator)

    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_callback(entry::CouplingCallback) -> Dict{String,Any}

Serialize callback coupling entry.
"""
function serialize_callback(entry::CouplingCallback)::Dict{String,Any}
    result = Dict{String,Any}("type" => "callback", "callback_id" => entry.callback_id)

    if entry.config !== nothing
        result["config"] = entry.config
    end
    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_event(entry::CouplingEvent) -> Dict{String,Any}

Serialize event coupling entry.
"""
function serialize_event(entry::CouplingEvent)::Dict{String,Any}
    result = Dict{String,Any}(
        "type" => "event",
        "event_type" => entry.event_type,
        "affects" => [serialize_affect_equation(a) for a in entry.affects]
    )

    if entry.conditions !== nothing
        result["conditions"] = [serialize_expression(c) for c in entry.conditions]
    end
    if entry.trigger !== nothing
        result["trigger"] = serialize_discrete_event_trigger(entry.trigger)
    end
    if entry.affect_neg !== nothing
        result["affect_neg"] = [serialize_affect_equation(a) for a in entry.affect_neg]
    end
    if entry.discrete_parameters !== nothing
        result["discrete_parameters"] = entry.discrete_parameters
    end
    if entry.root_find !== nothing
        result["root_find"] = entry.root_find
    end
    if entry.reinitialize !== nothing
        result["reinitialize"] = entry.reinitialize
    end
    if entry.description !== nothing
        result["description"] = entry.description
    end

    return result
end

"""
    serialize_reference(ref::Reference) -> Dict{String,Any}

Serialize Reference to JSON-compatible format.
"""
function serialize_reference(ref::Reference)::Dict{String,Any}
    result = Dict{String,Any}()
    if ref.doi !== nothing
        result["doi"] = ref.doi
    end
    if ref.citation !== nothing
        result["citation"] = ref.citation
    end
    if ref.url !== nothing
        result["url"] = ref.url
    end
    if ref.notes !== nothing
        result["notes"] = ref.notes
    end
    return result
end

"""
    serialize_metadata(metadata::Metadata) -> Dict{String,Any}

Serialize Metadata to JSON-compatible format.
"""
function serialize_metadata(metadata::Metadata)::Dict{String,Any}
    result = Dict{String,Any}("name" => metadata.name)

    if metadata.description !== nothing
        result["description"] = metadata.description
    end
    if !isempty(metadata.authors)
        result["authors"] = metadata.authors
    end
    if metadata.license !== nothing
        result["license"] = metadata.license
    end
    if metadata.created !== nothing
        result["created"] = metadata.created
    end
    if metadata.modified !== nothing
        result["modified"] = metadata.modified
    end
    if !isempty(metadata.tags)
        result["tags"] = metadata.tags
    end
    if !isempty(metadata.references)
        result["references"] = [serialize_reference(r) for r in metadata.references]
    end

    return result
end

"""
    serialize_domain(domain::Domain) -> Dict{String,Any}

Serialize Domain to JSON-compatible format.
"""
function serialize_domain(domain::Domain)::Dict{String,Any}
    result = Dict{String,Any}()
    if domain.spatial !== nothing
        result["spatial"] = domain.spatial
    end
    if domain.temporal !== nothing
        result["temporal"] = domain.temporal
    end
    return result
end

"""
    serialize_solver(solver::Solver) -> Dict{String,Any}

Serialize Solver to JSON-compatible format matching the ESM schema.
"""
function serialize_solver(solver::Solver)::Dict{String,Any}
    result = Dict{String,Any}("strategy" => solver_strategy_to_string(solver.strategy))

    # Serialize configuration
    config_dict = serialize_solver_configuration(solver.config)
    if !isempty(config_dict)
        result["config"] = config_dict
    end

    return result
end

"""
    serialize_solver_configuration(config::SolverConfiguration) -> Dict{String,Any}

Serialize SolverConfiguration to JSON-compatible format.
"""
function serialize_solver_configuration(config::SolverConfiguration)::Dict{String,Any}
    result = Dict{String,Any}()

    # Add basic configuration
    if config.threads !== nothing
        result["threads"] = config.threads
    end

    if config.timestep !== nothing
        result["timestep"] = config.timestep
    end

    # Add algorithm selections
    if config.stiff_algorithm !== nothing
        result["stiff_algorithm"] = config.stiff_algorithm
    end

    if config.nonstiff_algorithm !== nothing
        result["nonstiff_algorithm"] = config.nonstiff_algorithm
    end

    if config.map_algorithm !== nothing
        result["map_algorithm"] = config.map_algorithm
    end

    # Add stiff solver parameters
    if !isempty(config.stiff_kwargs)
        result["stiff_kwargs"] = config.stiff_kwargs
    end

    # Add numerical method if specified
    if config.numerical_method !== nothing
        result["numerical_method"] = numerical_method_to_string(config.numerical_method)
    end

    # Add extra parameters
    for (key, value) in config.extra_parameters
        if !haskey(result, key)  # Don't overwrite existing keys
            result[key] = value
        end
    end

    return result
end

"""
    serialize_esm_file(file::EsmFile) -> Dict{String,Any}

Serialize EsmFile to JSON-compatible format.
"""
function serialize_esm_file(file::EsmFile)::Dict{String,Any}
    result = Dict{String,Any}(
        "esm" => file.esm,
        "metadata" => serialize_metadata(file.metadata)
    )

    if file.models !== nothing
        result["models"] = Dict(k => serialize_model(v) for (k, v) in file.models)
    end
    if file.reaction_systems !== nothing
        result["reaction_systems"] = Dict(k => serialize_reaction_system(v) for (k, v) in file.reaction_systems)
    end
    if file.data_loaders !== nothing
        result["data_loaders"] = Dict(k => serialize_data_loader(v) for (k, v) in file.data_loaders)
    end
    if file.operators !== nothing
        result["operators"] = Dict(k => serialize_operator(v) for (k, v) in file.operators)
    end
    if !isempty(file.coupling)
        result["coupling"] = [serialize_coupling_entry(c) for c in file.coupling]
    end
    if file.domain !== nothing
        result["domain"] = serialize_domain(file.domain)
    end
    if file.solver !== nothing
        result["solver"] = serialize_solver(file.solver)
    end

    return result
end

"""
    save(file::EsmFile, path::String)

Save an EsmFile object to a JSON file at the specified path.
"""
function save(file::EsmFile, path::String)
    open(path, "w") do io
        save(file, io)
    end
end

"""
    save(file::EsmFile, io::IO)

Save an EsmFile object to a JSON stream.
"""
function save(file::EsmFile, io::IO)
    serialized = serialize_esm_file(file)
    write(io, JSON3.write(serialized, indent=2))
end